# data analysis with Python
 

In [ ]:
import numpy as np

In [3]:
l=[1,2,3,]
a =np.array([1,2,3])

In [4]:
for e in l:
    print(e)

1
2
3


In [5]:
 for e in a:
        print(e)

1
2
3


In [6]:
#comparing list and array and numpy
l.append(4)


 # We can see that append method is not present in numpy array we cannot add nuber to the array like this 

 a.append(4)
 
 ---------------------------------------------------------------------------
 AttributeError                            Traceback (most recent call last)
 <ipython-input-7-a39724ffcd65> in <module>
 ----> 1 a.append(4)

 AttributeError: 'numpy.ndarray' object has no attribute 'append'
<
    



In [8]:
l+=[5]

In [9]:
l

[1, 2, 3, 4, 5]

In [10]:
l2=[]
for e in l:
    l2.append(e+e)

In [11]:
l2

[2, 4, 6, 8, 10]

In [12]:
a+a

array([2, 4, 6])

In [13]:
3*a


array([3, 6, 9])

In [14]:
3*l

[1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5]

In [15]:
l2=[]
for e in l:
    l2.append(e**2)
    

In [16]:
l2

[1, 4, 9, 16, 25]

In [17]:
a**2

array([1, 4, 9])

In [18]:
  np.sqrt(a)

array([1.        , 1.41421356, 1.73205081])

In [19]:
np.log(a)

array([0.        , 0.69314718, 1.09861229])

In [20]:
np.exp(a)

array([ 2.71828183,  7.3890561 , 20.08553692])

In [21]:
#numpy for vector 

In [22]:
#array in numpy can be treated as vector 

In [23]:
a= np.array([1,2])
b= np.array([2,1])

In [24]:
dot = 0
for e, f in zip(a,b):
    dot+=e*f

In [25]:
dot

4

In [26]:
a*b

array([2, 2])

In [29]:
np.sum(a*b)

4

In [30]:
(a*b).sum()

4

In [31]:
np.dot(a,b)

4

In [32]:
a.dot(b)

4

In [33]:
b.dot(a)

4

In [34]:
#finding angle of the vector 

In [35]:
amag = np.sqrt((a*a).sum())

In [36]:
amag

2.23606797749979

In [37]:
amag =np.linalg.norm(a)

In [38]:
amag

2.23606797749979

In [40]:
cosangle =a.dot(b)/(np.linalg.norm(a)*np.linalg.norm(b))

In [41]:
cosangle


0.7999999999999998

In [42]:
angle = np.arccos(cosangle)

In [43]:
angle 

0.6435011087932847

In [44]:
#this angle is by default in radians 

In [50]:
#this program compares the time taken for computing dot product by different methods
import numpy as np 
from datetime import datetime
a = np.random.randn(100)
b = np.random.randn(100)
T= 100000
def slow_dot_product(a,b):
    result = 0
    for e,f in zip(a,b):
        result +=e*f
    return result

t0 = datetime.now()
for t in range(T):
    slow_dot_product(a,b)
dt1 = datetime.now() -t0

t0 =datetime.now()
for t in range(T):
    a.dot(b)
dt2 = datetime.now() -t0

print ("dt1/dt2",dt1.total_seconds()/dt2.total_seconds())

dt1/dt2 61.15585196607555


In [73]:
# we can see that the numpy is much faster than for loop in python

# vector and matices in numpy

In [53]:
M = np.array([[1,2],[3,4]])
l= [[1,2],[3,4]]

In [54]:
l[0]

[1, 2]

In [55]:
l[0][0]

1

In [56]:
M[0][0]

1

In [60]:
M[0,0]
#this is sanother notation that can be used 

1

In [70]:
M2 = np.matrix([[1,2],[3,4]])
M2

matrix([[1, 2],
        [3, 4]])

## this matrix in numpy works somewhat same as numpy array but it is not same 
most of the time we use numpy array 
and the ooficial documentaion of numpy advises us not to use matrix

In [72]:
#conerting matrix into array
A = np.array(M2)
A

array([[1, 2],
       [3, 4]])

## generating matrix to work with 

### one way to this is to pass list in the np.array()

In [74]:
np.array([1,2,3])

array([1, 2, 3])

this is inconvenient as We have to type each element individually

In [76]:
# array of Zeros 
Z = np.zeros(10)

In [77]:
Z

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [78]:
Z1 = np.zeros((10,10))

In [79]:
Z1

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])

In [82]:
#array of rando numbers 
Ones = np.ones((10,10))

In [83]:
Ones

array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

In [84]:
#random array
r = np.random.random((10,10))

In [85]:
r

array([[0.3988777 , 0.52856233, 0.70619845, 0.99089854, 0.27140298,
        0.65608355, 0.24680352, 0.49663351, 0.26305309, 0.93137392],
       [0.77836055, 0.38903653, 0.18425771, 0.49029553, 0.52602236,
        0.00969344, 0.27250648, 0.4772282 , 0.17238815, 0.51617804],
       [0.93578082, 0.71694937, 0.15887012, 0.54866109, 0.13370636,
        0.58111842, 0.41501452, 0.95844318, 0.82568879, 0.93986519],
       [0.51530416, 0.16089235, 0.90331623, 0.70847994, 0.04929376,
        0.56194766, 0.73232022, 0.76649099, 0.45255977, 0.76630804],
       [0.60129892, 0.3305913 , 0.40077531, 0.78774177, 0.94909447,
        0.71741496, 0.70867906, 0.43110999, 0.5899841 , 0.87143448],
       [0.02946294, 0.50891985, 0.27894144, 0.30483243, 0.21154913,
        0.10478531, 0.68532357, 0.76711738, 0.17142865, 0.47259541],
       [0.01275755, 0.67832167, 0.61969535, 0.85879291, 0.88388341,
        0.98892072, 0.43804492, 0.65129651, 0.56851178, 0.19335077],
       [0.42682397, 0.10422225, 0.2337401

In [86]:
# this random funtion use normal distribution

In [87]:
#gaussian distribution matrix
G = np.random.randn(10,10)

In [88]:
G

array([[ 1.09898139e+00, -6.97010752e-01,  3.03173395e-01,
        -2.00755689e-01,  5.69495294e-01,  4.59861465e-01,
        -6.48313891e-01, -5.13914256e-01, -2.83558829e+00,
         8.68317644e-01],
       [ 9.12335344e-02, -1.06144281e+00,  3.51641960e-01,
        -1.50175664e+00, -7.57018172e-01, -8.51622928e-01,
        -1.70647297e+00, -6.10920379e-02, -6.90658259e-02,
        -2.46217163e-01],
       [-5.90213763e-02, -2.96441593e-01, -7.41789279e-01,
        -2.40230295e-01, -2.64785050e-01, -1.15251856e+00,
         4.89409210e-01,  2.91927625e+00, -2.33754123e-01,
         6.62743996e-01],
       [ 9.99122770e-01,  1.05776395e+00,  1.52169636e+00,
         9.91839308e-01, -3.88950880e-01,  5.39315195e-01,
         1.81053839e+00, -9.27935358e-01,  2.26104607e-01,
         7.97437431e-01],
       [ 2.71495460e-03, -9.09289041e-01,  1.36795922e+00,
        -1.15378661e+00,  1.31081551e-01,  6.39632656e-01,
        -1.03709103e-01,  5.90744704e-01, -2.06499935e+00,
        -2.

In [89]:
G.mean()

-0.06503093691034328

In [90]:
G.var()

0.9772639927550664

# Matrix products 
  Matrix multipication
  requirement : inner dimentions must match
  if we have A of size (2,3) and B of size (3,3)
  we can multiply AB (inner dimention is 3)
  We cannot multiply BA (inner dimention is 3/2)
  
  </n> C(i,j) = summation of (A(i,k)B(k,j)) for k = 1 to K

In [93]:
a = np.array([[1,2,3],[4,5,6]])
b = np.array([[1,2,3],[4,5,6],[7,8,9]])

In [94]:
c = a.dot(b)

In [95]:
c

array([[30, 36, 42],
       [66, 81, 96]])

### inverse fuction 

In [96]:
A = np.array([[1,2],[3,4]])

In [97]:
Ainv = np.linalg.inv(A)

In [98]:
Ainv

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

In [100]:
Ainv.dot(A)

array([[1.0000000e+00, 4.4408921e-16],
       [0.0000000e+00, 1.0000000e+00]])

In [102]:
#determnant
np.linalg.det(A)

-2.0000000000000004

In [104]:
#diagonal elements
np.diag(A)

array([1, 4])

In [105]:
np.diag([1,2])

array([[1, 0],
       [0, 2]])

In [106]:
a = np.array([1,2])
b = np.array([3,4])

In [107]:
 np.outer(a,b)

array([[3, 4],
       [6, 8]])

In [108]:
np.inner(a,b)

11

In [109]:
a.dot(b)

11

In [110]:
a*b

array([3, 8])

In [112]:
#sum of diagonal 
np.diag(A).sum()

5

In [113]:
np.trace(A)

5

# Eigen values and Eigen vectors 

In [114]:
X = np.random.randn(100,3)

In [116]:
cov = np.cov(X)

In [117]:
cov.shape

(100, 100)

In [122]:
cov = np.cov(X.T)
#remember when calculating covarience we have to trnspose it first

In [123]:
cov

array([[ 0.7868759 ,  0.00448238,  0.06108376],
       [ 0.00448238,  0.97215503, -0.03819237],
       [ 0.06108376, -0.03819237,  1.00090849]])

# Eignevalues and Evectors 
</n> eigenvalues , eigen vectors  = np.eig(C)

</n>OR

</n>eigenvalues, eigenvectors  = np.eigh(C)

</n> eigh is for symmetric and Hermitian Matrices


In [124]:
np.linalg.eigh(cov)

(array([0.7696587 , 0.95322586, 1.03705486]),
 array([[ 0.96139642,  0.19009371,  0.19895049],
        [-0.07140202,  0.87059217, -0.4867967 ],
        [-0.26574173,  0.45379914,  0.85055727]]))

In [125]:
np.linalg.eigh(cov)

(array([0.7696587 , 0.95322586, 1.03705486]),
 array([[ 0.96139642,  0.19009371,  0.19895049],
        [-0.07140202,  0.87059217, -0.4867967 ],
        [-0.26574173,  0.45379914,  0.85055727]]))

# solving a Linear System 

In [126]:
A

array([[1, 2],
       [3, 4]])

In [129]:
b = np.array([1,2])
b

array([1, 2])

In [130]:
x = np.linalg.inv(A).dot(b)

In [131]:
x

array([2.22044605e-16, 5.00000000e-01])

In [132]:
# we can use solve in numpy to dorectly solve the eqaution

In [134]:
x = np.linalg.solve(A,b)

In [135]:
x 

array([0. , 0.5])

### Example Problem
The admission fee st a small fair is  1.50  dollar for children and 4.00 dollar for adults .On a certain day ,2200 people enter the fair and $ 5050 is collected . How many children and how many adults attended?

let :

x1 = no of chilren

x2 = no of adults

1.5x1 + 4x2 = 5050

In [137]:
A = np.array([[1,1],[1.5,4]])

In [138]:
b = np.array([2200, 5050]) 

In [139]:
np.linalg.solve(A,b)

array([1500.,  700.])